Optimize NCHW2NHWC on GPU

xiaomengy opened the pull request:
Summary: Optimize NCHW2NHWC on GPU

Differential Revision: D10481163

BIT-silence commented:
benchmark:
N = 128, C = 256, H = 152, W = 152
NCNW2NHWC: 401.25ms -> 35.58ms
NHWC2NCHW: 339.50ms -> 37.44ms

BIT-silence commented:
2D Transpose,
R = 10000, C = 10000,
58ms -> 5.7ms
CuDNN: 20.78ms

BIT-silence commented on caffe2/utils/math_gpu.cu:
```diff
@@ -3836,6 +3839,50 @@ CAFFE2_SPECIALIZED_CUDA_INV_STD(float)
+constexpr int kBlockRows = 8;
+
+template <typename T>
+__global__ void BatchTranspose2DCUDAKernel(
```
Done

xiaomengy committed:
Optimize NCHW2NHWC on GPU (#12910)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12910

Optimize NCHW2NHWC on GPU

Reviewed By: houseroad

Differential Revision: D10481163

fbshipit-source-id: da1f2ef3b5337a42f2d10929a55a2d3a9c82bdaa
